### ISSUE TITLE
Process objects created with Start-Process -NoNewWindow and/or -RedirectStandard* never report their exit code on Windows.

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#5421]
Note:

* The problem only surfaces on Windows.
* The problem does _not_ occur when you use `-Wait` in addition to `-PassThru`  as opposed to a separate `Wait-Process` call.
  * _Update_: Per #20400, the problem now (PowerShell 7.4.0-preview.5) also surfaces when `-Wait` _is_ used.    

**Workaround**:

Per https://github.com/PowerShell/PowerShell/issues/20716#issuecomment-1817967688, the workaround is to cache the process handle before calling `Wait-Process` / `.WaitForExit()`:

```powershell
$p = Start-Process cmd -args '/c', 'ver' -PassThru -NoNewWindow
$dummy = $p.Handle # WORKAROUND: use a dummy variable to cache the handle
Wait-Process -id $p.Id
"Exit code: [$($p.ExitCode)]"  # Now the exit code is reported.
```

Steps to reproduce
------------------

On Windows:

```powershell
$p = Start-Process cmd -args '/c', 'ver' -PassThru -NoNewWindow
Wait-Process -id $p.Id
"Exit code: [$($p.ExitCode)]"
```

Note that the problem only occurs when `-NoNewWindows` and /or any of the `-RedirectStandard*` parameters are present.

Inserting a `$p.WaitForExit()` call after the `Wait-Process` call, as suggested [in the docs](https://msdn.microsoft.com/en-us/library/system.diagnostics.process.exitcode%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) to ensure that `.ExitCode` has a value (which should be the equivalent of `Wait-Process`), doesn't help.

Expected behavior
-----------------

```none

Microsoft Windows [Version 10.0.15063]
Exit code: [0]
```

`$ps.ExitCode` should contain `0`, given that the `cmd` command reports exit code `0`.

Actual behavior
---------------

```none

Microsoft Windows [Version 10.0.15063]
Exit code: []
```

Note how `$ps.ExitCode` is unexpectedly stringified to the empty string.
Also, even though `Get-Member` indicates that the property's date type is `[int]`, `$null -eq $ps.ExitCode` is `$true`.

This suggests that an exception is occurring behind the scenes, which PowerShell quietly ignores.

An exception when accessing `.ExitCode` should only occur if the process hasn't exited yet, however, which is at odds with having used `Wait-Process` and `$p.HasExited` indicating `$true`.

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.9 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.674 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

**Comment by vexx32** [comment]
https://stackoverflow.com/questions/10262231/obtaining-exitcode-using-start-process-and-waitforexit-instead-of-wait

Dug this up... Seems as though it's not so much waiting for it to end as it is *explicitly querying for the process handle* while it's still running is what allows the exit code to be captured. 

Ideally, `Start-Process` could internally query the process handle from the object before passing it to the pipeline in order to ensure the exit code is always captured, regardless of the circumstance.

**Comment by mklement0** [comment]
Great find, @vexx32 - that sounds like a great workaround.

A little more background info to flesh out your statement from a [linked answer](https://stackoverflow.com/a/23797762/45375)

> This is a quirk of the implementation of the .NET Process object. The implementation of the ExitCode property first checks if the process has exited. 
For some reason, the code that performs that check not only looks at the HasExited property but also [verifies that the process handle is present in the process object and throws an exception if it is not](https://source.dot.net/#System.Diagnostics.Process/System/Diagnostics/Process.cs,954). PowerShell intercepts that exception and returns null. 
Accessing the Handle property causes the process object to retrieve the process handle and store it internally. Once the handle is stored in the process object, the ExitCode property works as expected.

I wonder if there was ever a good reason for this behavior, or whether we should also file a CoreFx bug report.

**Comment by jmaxxz** [comment]
Any updates on this? This really should be addressed.

**Comment by jordanrh1** [comment]
Any update? This is a rather impactful productivity blocker because you can't tell if process succeeded or failed. Makes start-process pretty much useless.

**Comment by mklement0** [comment]
Per https://github.com/PowerShell/PowerShell/issues/20716#issuecomment-1817967688, the workaround (only applicable if `-Wait` isn't used)  is to cache the process handle before calling `Wait-Process` / `.WaitForExit()`; I've also updated the initial post with an example.

**Comment by mklement0** [comment]
A fix _may_ land in  v7.4.1 (the PR mentions only the related #20400 issue):

* #20866

**Comment by mklement0** [comment]
It looks like the problem has indeed been fixed, as verified via the original repro steps, in the current version, v7.4.2.
